File: src/utils/MoveGenerator.java
package utils;

import game.Board;
import game.Move;
import game.Piece;
import game.PrecomputedGameData;

import java.io.PipedInputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class MoveGenerator {
    public static List<Move> generateMoves(Board board) {
        if (board.isGameOver()) {
            return new ArrayList<>();
        }

        List<Move> moves = new ArrayList<>();
        int color = board.getTurn();
        List<Integer> piecePositions = board.getPiecePositions(color);

        for (int index : piecePositions) {
            if (index == -1) {
                continue;
            }

            switch (board.getPieceType(index)) {
                case Piece.PAWN:
                    List<Move> pawnMoves = generatePawnMoves(board, index);
                    if (pawnMoves.size() > 0) {
                        moves.addAll(pawnMoves);
                    }
                    break;
                case Piece.BISHOP:
                    List<Move> bishopMoves = generateBishopMoves(board, index);
                    if (bishopMoves.size() > 0) {
                        moves.addAll(bishopMoves);
                    }
                    break;
                case Piece.ROOK:
                    List<Move> rookMoves = generateRookMoves(board, index);
                    if (rookMoves.size() > 0) {
                        moves.addAll(rookMoves);
                    }
                    break;
                case Piece.QUEEN:
                    List<Move> queenMoves = generateQueenMoves(board, index);
                    if (queenMoves.size() > 0) {
                        moves.addAll(queenMoves);
                    }
                    break;
                case Piece.KING:
                    List<Move> kingMoves = generateKingMoves(board, index);
                    kingMoves.addAll(checkCastling(board, index));
                    if (kingMoves.size() > 0) {
                        moves.addAll(kingMoves);
                    }
                    break;
                case Piece.KNIGHT:
                    List<Move> knightMoves = generateKnightMoves(board, index);
                    if (knightMoves.size() > 0) {
                        moves.addAll(knightMoves);
                    }
                    break;
            }
        }
        return moves;
    }

    public static boolean legalMovesExist(Board board) {
        int color = board.getTurn();
        List<Integer> piecePositions = board.getPiecePositions(color);

        for (int index : piecePositions) {
            if (index == -1) {
                continue;
            }

            switch (board.getPieceType(index)) {
                case Piece.PAWN:
                    List<Move> pawnMoves = generatePawnMoves(board, index);
                    if (pawnMoves.size() > 0) {
                        return true;
                    }
                    break;
                case Piece.BISHOP:
                    List<Move> bishopMoves = generateBishopMoves(board, index);
                    if (bishopMoves.size() > 0) {
                        return true;
                    }
                    break;
                case Piece.ROOK:
                    List<Move> rookMoves = generateRookMoves(board, index);
                    if (rookMoves.size() > 0) {
                        return true;
                    }
                    break;
                case Piece.QUEEN:
                    List<Move> queenMoves = generateQueenMoves(board, index);
                    if (queenMoves.size() > 0) {
                        return true;
                    }
                    break;
                case Piece.KING:
                    List<Move> kingMoves = generateKingMoves(board, index);
                    kingMoves.addAll(checkCastling(board, index));
                    if (kingMoves.size() > 0) {
                        return true;
                    }
                    break;
                case Piece.KNIGHT:
                    List<Move> knightMoves = generateKnightMoves(board, index);
                    if (knightMoves.size() > 0) {
                        return true;
                    }
                    break;
            }
        }

        return false;
    }

    private static List<Move> generatePawnMoves(Board board, int index) {
        List<Move> moves = new ArrayList<>();
        int color = board.getPieceColor(index);
        int[] dir = PrecomputedGameData.pieceDirections.get(Piece.create(Piece.PAWN, color));

        //if space in direction is free and move is legal, add move
        if (PrecomputedGameData.edgeOfBoard.get(dir[0])[index] > 0 && board.isEmpty(index + dir[0]) && board.isLegalMove(index, index + dir[0])) {

            //if a pawn can only move once before hitting the edge, he is going to promote
            if (PrecomputedGameData.edgeOfBoard.get(dir[0])[index] == 1) {
                moves.add(new Move(index, index + dir[0], 3, Piece.QUEEN));
                moves.add(new Move(index, index + dir[0], 3, Piece.ROOK));
                moves.add(new Move(index, index + dir[0], 3, Piece.BISHOP));
                moves.add(new Move(index, index + dir[0], 3, Piece.KNIGHT));
            } else {
                moves.add(new Move(index, index + dir[0], 0, 0));
            }
        }

        //if pawn can go forward 6 times it has not moved yet, check if moving 2 tiles is possible and if the tile in front of you is empty
        if (PrecomputedGameData.edgeOfBoard.get(dir[0])[index] == 6 && board.isEmpty(index + dir[0]) && board.isEmpty(index + (dir[0] * 2)) && board.isLegalMove(index, index + (dir[0] * 2))) {
            moves.add(new Move(index, index + (dir[0] * 2), 0, 0));
        }

        //look sideways for captures
        for (int side = 1; side < dir.length; side++) {
            if (PrecomputedGameData.edgeOfBoard.get(dir[side])[index] == 0) {
                continue;
            }

            //check for regular capture
            if (!board.isEmpty(index + dir[side]) && !board.isColor(index + dir[side], color) && board.isLegalMove(index, index + dir[side])) {
                //if a pawn can only move forward once before hitting the edge, he is going to promote
                if (PrecomputedGameData.edgeOfBoard.get(dir[0])[index] == 1) {
                    moves.add(new Move(index, index + dir[side], 3, Piece.QUEEN));
                    moves.add(new Move(index, index + dir[side], 3, Piece.ROOK));
                    moves.add(new Move(index, index + dir[side], 3, Piece.BISHOP));
                    moves.add(new Move(index, index + dir[side], 3, Piece.KNIGHT));
                } else {
                    moves.add(new Move(index, index + dir[side], 0, 0));
                }
            }

            //from this point on we check for en passant, don't waste time if it's not possible
            if (board.getEnPassant() == 0) {
                continue;
            }

            //if we get here, en passant is possible on the board, so check if this pawn can take it
            int enPassantCapt = board.getEnPassant() + dir[0];
            if (enPassantCapt == index + dir[side] && board.isLegalMove(index, index + dir[side])) {
                moves.add(new Move(index, index + dir[side], 2, 0));
            }
        }

        return moves;
    }

    private static List<Move> generateBishopMoves(Board board, int index) {
        return scanDirectionUntilCollision(board, index, PrecomputedGameData.pieceDirections.get(Piece.BISHOP));
    }

    private static List<Move> generateKnightMoves(Board board, int index) {
        return scanDirectionOnce(board, index, PrecomputedGameData.pieceDirections.get(Piece.KNIGHT));
    }

    private static List<Move> generateRookMoves(Board board, int index) {
        return scanDirectionUntilCollision(board, index, PrecomputedGameData.pieceDirections.get(Piece.ROOK));
    }

    private static List<Move> generateQueenMoves(Board board, int index) {
        List<Move> moves = new ArrayList<>();
        moves.addAll(generateRookMoves(board, index));
        moves.addAll(generateBishopMoves(board, index));
        return moves;
    }

    private static List<Move> generateKingMoves(Board board, int index) {
        return scanDirectionOnce(board, index, PrecomputedGameData.pieceDirections.get(Piece.KING));
    }

    private static List<Move> checkCastling(Board board, int index) {
        List<Move> castleMove = new ArrayList<>();
        int color = board.getPieceColor(index);

        if ((!board.canCastleKingSide(color) && !board.canCastleQueenSide(color)) || board.isCheck()) {
            return castleMove;
        }

        //if the tiles between king and rook are empty
        if (board.isEmpty(index + 1) && board.isEmpty(index + 2) && board.canCastleKingSide(color)) {
            //if there is a piece at the end, and it is your rook, which hasn't moved
            if (!board.isEmpty(index + 3) && board.isRook(index + 3) && board.isColor(index + 3, color)) {
                //if move is legal
                if (board.isLegalMove(index, index + 1) && board.isLegalMove(index, index + 2)) {
                    castleMove.add(new Move(index, index + 2, 1, 0));
                }
            }
        }

        //if the tiles between king and rook are empty
        if (board.isEmpty(index - 1) && board.isEmpty(index - 2) && board.isEmpty(index - 3) && board.canCastleQueenSide(color)) {
            //if there is a piece at the end, and it is your rook, which hasn't moved
            if (!board.isEmpty(index - 4) && board.isRook(index - 4) && board.isColor(index - 4, color)) {
                //if move is legal
                if (board.isLegalMove(index, index - 1) && board.isLegalMove(index, index - 2)) {
                    castleMove.add(new Move(index, index - 2, 1, 0));
                }
            }
        }

        return castleMove;
    }

    private static List<Move> scanDirectionOnce(Board board, int index, int[] directions) {
        List<Move> moves = new ArrayList<>(8);
        int color = board.getPieceColor(index);

        for (int direction : directions) {
            int count = PrecomputedGameData.edgeOfBoard.get(direction)[index];

            if (count > 0 && (board.isEmpty(index + direction) || !board.isColor(index + direction, color))) {
                if (board.isLegalMove(index, index + direction)) {
                    moves.add(new Move(index, index + direction, 0, 0));
                }
            }
        }

        return moves;
    }

    private static List<Move> scanDirectionUntilCollision(Board board, int index, int[] directions) {
        List<Move> moves = new ArrayList<>(8);
        int color = board.getPieceColor(index);

        for (int direction : directions) {
            int count = PrecomputedGameData.edgeOfBoard.get(direction)[index];
            int curTile = index;

            while (count > 0 && (board.isEmpty(curTile + direction) || !board.isColor(curTile + direction, color))) {
                curTile += direction;

                if (board.isLegalMove(index, curTile)) {
                    moves.add(new Move(index, curTile, 0, 0));
                }

                if (!board.isEmpty(curTile) && !board.isColor(curTile, color)) {
                    break;
                }
                count--;
            }
        }

        return moves;
    }
}

File: src/moveGenerator.cpp
#include "moveGenerator.h"

#include <bit>
#include <unordered_map>

#include "move.h"
#include "piece.h"
#include "util/precomputedGameData.h"
#include "util/utils.h"

// <--> MOVE GENERATION <--> //

std::vector<Move> MoveGenerator::generateMoves(Board& board) {
    std::vector<Move> moves;
    moves.reserve(218);

    if (board.isGameOver()) {
        return moves;
    }

    const int turn = board.getTurn();
    uint64_t piecePositions = board.getPiecePositionsColor(turn);

    while (piecePositions) {
        const int piecePosition = popLSB(piecePositions);

        switch (board.getPieceType(piecePosition)) {
            case Piece::PAWN: {
                generatePawnMoves(board, piecePosition, moves);
                break;
            }
            case Piece::BISHOP: {
                generateSlidingPieceMoves(board, piecePosition, Piece::BISHOP, moves);
                break;
            }
            case Piece::ROOK: {
                generateSlidingPieceMoves(board, piecePosition, Piece::ROOK, moves);
                break;
            }
            case Piece::QUEEN: {
                generateSlidingPieceMoves(board, piecePosition, Piece::BISHOP, moves);
                generateSlidingPieceMoves(board, piecePosition, Piece::ROOK, moves);
                break;
            }
            case Piece::KING: {
                generateKingMoves(board, piecePosition, moves);
                generateCastleMoves(board, piecePosition, moves);
                break;
            }
            case Piece::KNIGHT: {
                generateKnightMoves(board, piecePosition, moves);
                break;
            }
        }
    }

    return moves;
}

void MoveGenerator::generatePawnMoves(Board& board, const int index, std::vector<Move> &moves, bool returnEarly) {
    //double check, only the king can move
    if (std::popcount(board.getCheckers()) > 1) {
        return;
    }

    const int color = board.getPieceColor(index);
    const std::vector<int>& dir = PGD::getPieceDirections(Piece::create(Piece::PAWN, color));

    //NOTE: It is impossible for a pawn to not be able to move forward, given that there isn't a piece blocking it.
    //      This is because if a pawn gets to the edge of the board, it promotes

    const uint64_t friendly = board.getPiecePositionsColor(board.getTurn());
    const uint64_t enemy = board.getPiecePositionsColor(abs(board.getTurn() - 1));
    const uint64_t allPieces = friendly | enemy;

    //check if push is allowed
    uint64_t movesBB = (1LL << index + dir[0]) & ~allPieces;

    const int startRank = (board.getTurn() == Piece::WHITE) ? 1 : 6;
    const uint64_t startRankMask = rankMask(startRank);
    const uint64_t isOnStartRank = (1LL << index) & startRankMask;

    //check if double push is allowed
    if (movesBB && isOnStartRank > 0) {
        uint64_t doublePush = (1LL << index + (dir[0] * 2)) & ~allPieces;

        if (board.isCheck())
            doublePush &= board.getCheck();
        if (board.getPins(index, board.getTurn()))
            doublePush &= board.getPins(index, board.getTurn());

        if (doublePush)
            moves.emplace_back(Move(index, index + (dir[0] * 2), Flag::DBL_PUSH));
    }

    //check captures
    const uint64_t attackMap = PGD::getAttackMap(Piece::create(Piece::PAWN, color), index);
    const uint64_t capturesBB = attackMap & enemy;
    movesBB |= capturesBB;

    //this is true if the en passant pawn is right next to us
    int enPassant = board.getEnPassant();
    if (enPassant == index - 1 || enPassant == index + 1) {

        //we get the en passant direction like this because the condition is the other way around for black
        //and I didn't want to write another if statement
        int epDir = enPassant > index ? dir[2 - board.getTurn()] : dir[1 + board.getTurn()];
        int end = index + epDir;

        //check that the pawn doesn't wrap around the board to take en passant
        if ((attackMap & (1LL << end)) != 0 && board.isLegalMove(index, end, Flag::EN_PASSANT))
            moves.emplace_back(Move(index, end, Flag::EN_PASSANT));
    }

    //pawn can only move in the check line or pin line
    if (board.isCheck())
        movesBB &= board.getCheck();
    if (board.getPins(index, board.getTurn()))
        movesBB &= board.getPins(index, board.getTurn());

    int promotionRank = board.getTurn() == Piece::WHITE ? 7 : 0;
    uint64_t promotionRankMask = rankMask(promotionRank);

    while (movesBB) {
        const int to = popLSB(movesBB);

        if ((promotionRankMask & (1LL << to)) > 0) {
            moves.emplace_back(index, to, Flag::PROMO_B);
            moves.emplace_back(index, to, Flag::PROMO_R);
            moves.emplace_back(index, to, Flag::PROMO_N);
            moves.emplace_back(index, to, Flag::PROMO_Q);
            continue;
        }

        Flag flag = (((1LL << to) & enemy) > 0) ? Flag::CAPTURE : Flag::QUIET;
        moves.emplace_back(index, to, flag);

        if (returnEarly)
            return;
    }
}

void MoveGenerator::generateKnightMoves(Board& board, const int index, std::vector<Move> &moves, bool returnEarly) {
    const uint64_t pseudoMoves = PGD::getAttackMap(Piece::create(Piece::KNIGHT, Piece::WHITE), index);
    uint64_t pseudoLegalMoves = pseudoMoves & ~board.getPiecePositionsColor(board.getPieceColor(index));

    //if it's a double check, only king moves allowed
    if (std::popcount(board.getCheckers()) > 1)
        return;

    //if it's check, you are only allowed to move on the checkline
    if (std::popcount(board.getCheckers()) == 1)
        pseudoLegalMoves &= board.getCheck();

    //if there is a pin, only moves along that pin line are legal
    //pinned pieces will not be able to move in checks
    const uint64_t pin = board.getPins(index, board.getTurn());
    if (pin)
        pseudoLegalMoves &= pin;

    while (pseudoLegalMoves) {
        const int moveTo = popLSB(pseudoLegalMoves);

        //check if we need to set the capture flag
        const Flag flag = (board.isEmpty(moveTo)) ? Flag::QUIET : Flag::CAPTURE;
        moves.emplace_back(index, moveTo, flag);

        if (returnEarly)
            return;
    }
}

void MoveGenerator::generateSlidingPieceMoves(Board& board, const int index, const int pieceType,
                                              std::vector<Move> &moves, bool returnEarly) {
    const uint64_t blockerBitboard = board.getPiecePositionsColor(Piece::WHITE) | board.
                                     getPiecePositionsColor(Piece::BLACK);

    //where we can move according to current blockers
    const uint64_t pseudoMoves = PGD::getPseudoMoves(pieceType, index, blockerBitboard);

    //take out friendly piece captures
    uint64_t pseudoLegalMoves = pseudoMoves & ~board.getPiecePositionsColor(board.getPieceColor(index));

    //if it's a double check, only king moves allowed
    if (std::popcount(board.getCheckers()) > 1)
        return;

    //if it's check, you are only allowed to move on the checkline
    if (std::popcount(board.getCheckers()) == 1)
        pseudoLegalMoves &= board.getCheck();

    //if there is a pin, only moves along that pin line are legal
    //pinned pieces will not be able to move in checks
    const uint64_t pin = board.getPins(index, board.getTurn());
    if (pin)
        pseudoLegalMoves &= pin;

    while (pseudoLegalMoves) {
        const int moveTo = popLSB(pseudoLegalMoves);

        //check if we need to set the capture flag
        const Flag flag = (board.isEmpty(moveTo)) ? Flag::QUIET : Flag::CAPTURE;
        moves.emplace_back(index, moveTo, flag);

        if (returnEarly)
            return;
    }
}

void MoveGenerator::generateKingMoves(Board& board, const int index, std::vector<Move> &moves, bool returnEarly) {
    const uint64_t pseudoMoves = PGD::getAttackMap(Piece::create(Piece::KING, Piece::WHITE), index);
    uint64_t pseudoLegalMoves = pseudoMoves & ~board.getPiecePositionsColor(board.getPieceColor(index));

    if (board.isCheck()) {
        //this is all for extending the check line
        uint64_t checkers = board.getCheckers();
        while (checkers) {
            int checkerIndex = popLSB(checkers);

            switch (board.getPieceType(checkerIndex)) {
                case Piece::KING:
                case Piece::KNIGHT:
                    pseudoLegalMoves &= ~PGD::getAttackMap(Piece::create(board.getPieceType(checkerIndex), Piece::WHITE), checkerIndex);
                break;
               case Piece::PAWN:
                    pseudoLegalMoves &= ~PGD::getAttackMap(Piece::create(Piece::PAWN, abs(board.getTurn() - 1)), checkerIndex);
                break;
                default:
                    //sliding pieces
                        uint64_t blockerBB = board.getPiecePositionsColor(Piece::WHITE) | board.getPiecePositionsColor(Piece::BLACK);
                //remove king from blockers
                blockerBB &= ~(1LL << index);
                uint64_t pieceView = PGD::getPseudoMoves(board.getPieceType(checkerIndex), checkerIndex, blockerBB);
                pseudoLegalMoves &= ~(pieceView);
                break;
            }
        }
    }

    uint64_t legalMoves = pseudoLegalMoves & ~board.getColorAttackMap(abs(board.getTurn() - 1));

    while (legalMoves) {
        const int moveTo = popLSB(legalMoves);

        //check if we need to set the capture flag
        const Flag flag = (board.isEmpty(moveTo)) ? Flag::QUIET : Flag::CAPTURE;
        moves.emplace_back(index, moveTo, flag);

        if (returnEarly)
            return;
    }
}

void MoveGenerator::generateCastleMoves(Board& board, const int index, std::vector<Move> &moves, bool returnEarly) {
    const int color = board.getTurn();

    if (board.isCheck()) {
        return;
    }

    const uint64_t friendly = board.getPiecePositionsColor(board.getTurn());
    const uint64_t enemy = board.getPiecePositionsColor(abs(board.getTurn() - 1));

    const uint64_t allPieces = friendly | enemy;
    const uint64_t enemyAttackMap = board.getColorAttackMap(abs(board.getTurn() - 1));

    if (board.canCastleKSide(color)) {
        uint64_t between = betweenMask(index, index + 3);

        //if the king would not be attacked while castling and there is no piece between king and rook
        if ((between & enemyAttackMap) == 0 && (between & allPieces) == 0)
            moves.emplace_back(index, index + 2, Flag::CASTLE_K);
    }

    if (board.canCastleQSide(color)) {
        uint64_t between = betweenMask(index, index - 4);

        //no pieces allowed between king and rook
        if ((between & allPieces) != 0) {
            return;
        }

        //index - 3 can be a check, it's fine
        between &= ~(1LL << index - 3);

        //if the king would not be attacked while castling, it's ok
        if ((between & enemyAttackMap) == 0)
            moves.emplace_back(index, index - 2, Flag::CASTLE_Q);
    }
}

// <--> MOVE GENERATION <--> //

// <--> CHECKING IF LEGAL MOVES EXIST <--> //

bool MoveGenerator::legalMovesExist(Board& board) {
    std::vector<Move> moves;

    if (board.isGameOver()) {
        return false;
    }

    const int turn = board.getTurn();
    uint64_t piecePositions = board.getPiecePositionsColor(turn);

    while (piecePositions) {
        const int piecePosition = popLSB(piecePositions);

        switch (board.getPieceType(piecePosition)) {
            case Piece::PAWN: {
                generatePawnMoves(board, piecePosition, moves, true);
                if (!moves.empty())
                    return true;
                break;
            }
            case Piece::BISHOP: {
                generateSlidingPieceMoves(
                        board, piecePosition, Piece::BISHOP, moves, true);
                if (!moves.empty())
                    return true;
                break;
            }
            case Piece::ROOK: {
                generateSlidingPieceMoves(board, piecePosition, Piece::ROOK, moves, true);
                if (!moves.empty())
                    return true;
                break;
            }
            case Piece::QUEEN: {
                generateSlidingPieceMoves(
                        board, piecePosition, Piece::BISHOP, moves, true);
                generateSlidingPieceMoves(
                        board, piecePosition, Piece::ROOK, moves, true);
                if (!moves.empty())
                    return true;
                break;
            }
            case Piece::KING: {
                generateKingMoves(board, piecePosition, moves, true);
                generateCastleMoves(board, piecePosition, moves, true);
                if (!moves.empty())
                    return true;
                break;
            }
            case Piece::KNIGHT: {
                generateKnightMoves(board, piecePosition, moves, true);
                if (!moves.empty())
                    return true;
                break;
            }
        }
    }

    return false;
}
